docs/tracing.md

```diff
@@ -0,0 +1,74 @@
+# Tracing
+
+## Introduction
+
+Instrumentation based tracing is implemented with
+[`log`](https://github.com/rust-lang/log) and
+[`log_instrument`](https://github.com/JonathanWoollett-Light/log-instrument),
+outputting a `Trace` level log when entering and exiting every function.
+
+It is disabled by default at compile-time, as such the
+**tracing functionality has no impact on the release binary.**
+
+To enable tracing, remove `"max_level_debug"` from the features
+for `log` under [`../src/vmm/Cargo.toml`](../src/vmm/Cargo.toml):
+
+```diff
+- log = { version = "0.4.17", features = ["std", "serde", "max_level_debug"] }
++ log = { version = "0.4.17", features = ["std", "serde"] }
+```
+
+This will result in a massive increase in the binary size and a
+massive regression in performance. To mitigate the performance affects
+you can use filtering.
+
+## Filtering
+
```

Comment:
I'd put a short introduction into the two filtering options at the beginning here, something like 

"Firecracker supports filtering tracing output both at run- and compile-time. This can be used to partially mitigate the performance impact of logging thousands of traces. The former allows you to enable tracing on a per-module via API-calls. The latter statically removes tracing from specific module, potentially resulting in better performance at the cost of decreased flexibility."

Reply:
I've added the text
```
You can filter tracing output both at run-time and compile-time.
This can be used to partially mitigate the performance impact of logging
many traces.

Run-time filtering is implemented with the `/logger` API call, this can
significantly mitigate the performance impact but cannot mitigate the
memory impact.

Compile-time filtering is a manual process assisted by the
[`clippy-tracing`](https://github.com/JonathanWoollett-Light/clippy-tracing)
tool, this can almost-entirely mitigate both the performance and memory
impact.
```

Reply:
Mhh, "mitigate the performance impact" is repeated quite a few times here. Can we say "performance impact" in the first paragraph, "execution time" in the second, and "execution time and memory" in the third?

Reply:
What about
```
You can filter tracing output both at run-time and compile-time.

Run-time filtering is implemented with the `/logger` API call, this can
significantly mitigate the performance impact but cannot mitigate the
memory impact.

Compile-time filtering is a manual process assisted by the
[`clippy-tracing`](https://github.com/JonathanWoollett-Light/clippy-tracing)
tool, this can almost-entirely mitigate both the performance and memory
impact.
```

Reply:
I've reworded it
```
You can filter tracing output both at run-time and compile-time.
This can be used to mitigate the performance impact of logging many
traces.

Run-time filtering is implemented with the `/logger` API call, this can
significantly mitigate the impact on execution time but cannot mitigate
the impact on memory usage.

Compile-time filtering is a manual process assisted by the
[`clippy-tracing`](https://github.com/JonathanWoollett-Light/clippy-tracing)
tool, this can almost-entirely mitigate the impact on execution time and
the impact on memory usage.
```